compiler/rustc_mir_build/src/build/expr/as_place.rs

```diff
@@ -316,6 +316,64 @@ impl<'tcx> PlaceBuilder<'tcx> {
         self.projection.push(elem);
         self
     }
+
+    pub fn try_ty<D>(&self, local_decls: &D, cx: &Builder<'_, 'tcx>) -> Option<Ty<'tcx>>
+    where
+        D: HasLocalDecls<'tcx>,
+    {
+        let tcx = cx.tcx;
+
+        let project_ty = |ty: Ty<'tcx>, elem: &PlaceElem<'tcx>| -> Ty<'tcx> {
+            match elem {
+                ProjectionElem::Deref => {
+                    ty.builtin_deref(true)
+                        .unwrap_or_else(|| {
+                            bug!("deref projection of non-dereferenceable ty {:?}", ty)
+                        })
+                        .ty
+                }
+                ProjectionElem::Index(_) | ProjectionElem::ConstantIndex { .. } => {
+                    ty.builtin_index().unwrap()
+                }
+                ProjectionElem::Subslice { from, to, from_end } => match ty.kind() {
+                    ty::Slice(..) => ty,
+                    ty::Array(inner, _) if !from_end => tcx.mk_array(*inner, (to - from) as u64),
+                    ty::Array(inner, size) if *from_end => {
+                        let size = size.eval_usize(tcx, ty::ParamEnv::empty());
+                        let len = size - (*from as u64) - (*to as u64);
+                        tcx.mk_array(*inner, len)
+                    }
+                    _ => bug!("cannot subslice non-array type: `{:?}`", ty),
+                },
+                ProjectionElem::Downcast(..) => ty,
+                ProjectionElem::Field(_, ty) | ProjectionElem::OpaqueCast(ty) => *ty,
+            }
+        };
+
+        match self.base {
+            PlaceBase::Local(local) => {
+                let base_ty = local_decls.local_decls()[local].ty;
+                Some(self.projection.iter().fold(base_ty, |ty, &elem| project_ty(ty, &elem)))
+            }
+            PlaceBase::Upvar { .. } => {
+                match to_upvars_resolved_place_builder(self.clone(), cx) {
+                    Ok(resolved_place_builder) => {
+                        // `base` is guaranteed to be `PlaceBase::Local` now, so recursive call is ok
+                        resolved_place_builder.try_ty(local_decls, cx)
+                    }
+                    Err(place_builder) => {
+                        match &place_builder.projection[..] {
+                            &[ProjectionElem::OpaqueCast(base_ty), ref projections @ ..] => Some(
```

Comment:
that's clever but I am not sure whether it is necessary :thinking: 

afaict the only reason why this returns `Err` is if the upvar is not part of the final mir body so it doesn't affect mir borrowck.

I would prefer always returning `None` in the `Err` case i think

Reply:
This is necessary for the following scenario (compiling with `--edition==2021`:

```
fn test1() {
    let t = (String::from("Hello"), String::from("bla"));

    let c = ||  {
        let (t1, t2) = t; 
        //~^ WARN unused variable: `t1`
        //~| WARN unused variable: `t2`

    };

    c();
}
```

where `y` is not captured as an upvar and we only get an `OpaqueCast` projection in the `PlaceBuilder`.

I'm not sure what the correct behaviour here is. One could argue that the name of the method implies a best-effort approach, so that we get the current type of the `PlaceBuilder` whenever possible, in which case these additional projections would be necessary.